File: 2008/S081.cpp
#include <cstdio>
int maxyears[1000001];
int dows[1000001];
int main() {
  int months[101];
  int m, w, l, mleap, dleap;
  int yul, mul, dul, a, b, n, sum = 0;
  scanf("%d %d %d %d %d", &m, &w, &l, &mleap, &dleap);
  for (int i = 1; i <= m; i++) {
    scanf("%d", &months[i]);
    sum += months[i];
  }
  scanf("%d %d %d %d %d", &yul, &mul, &dul, &a, &b);
  
  // first we assume the day before 1st Jan, 1 is day 0
  int dow = 0;
  for (int i = 1; i <= 1000000; i++) {
    dows[i] = dow;
    dow += sum + (i % l == 0 ? dleap : 0);
    dow %= w;
  }

  // find dow of given date
  n = dows[yul];
  for (int i = 1; i < mul; i++) {
    n += months[i] + (yul % l == 0 && i == mleap ? dleap : 0);
  }
  n = (n + dul) % w;

  // then we simulate with all different day of week on day before 1st Jan
  int noleap[100], leap[100];
  for (int i = 0; i < w; i++) {
    noleap[i] = 0;
    dow = i;
    for (int j = 1; j <= m; j++) {
      if (months[j] >= 13 && (dow + 13) % w == n) {
        noleap[i]++;
      }
      dow += months[j];
    }
  }
  for (int i = 0; i < w; i++) {
    leap[i] = 0;
    dow = i;
    for (int j = 1; j <= m; j++) {
      if (months[j] + (j == mleap ? dleap : 0) >= 13 && (dow + 13) % w == n) {
        leap[i]++;
      }
      dow += months[j] + (j == mleap ? dleap : 0);
    }
  }
  
  // lastly scan through year a to b to find max
  int max = 0;
  int maxcount = 0;
  for (int i = a; i <= b; i++) {
    if (i % l == 0) {
      if (leap[dows[i]] > max) {
        max = leap[dows[i]];
        maxyears[0] = i;
        maxcount = 1;
      } else if (leap[dows[i]] == max) {
        maxyears[maxcount] = i;
        maxcount++;
      }
    } else if (noleap[dows[i]] > max) {
      max = noleap[dows[i]];
      maxyears[0] = i;
      maxcount = 1;
    } else if (noleap[dows[i]] == max) {
      maxyears[maxcount] = i;
      maxcount++;
    }
  }
  
  // output
  printf("%d\n", max);
  for (int i = 0; i < maxcount - 1; i++) {
    printf("%d ", maxyears[i]);
  }
  printf("%d\n", maxyears[maxcount - 1]);
  return 0;
}

File: 2008/S081.pas
var
  maxyears, dows: array[1..1000000] of longint;
  months, noleap, leap: array[0..100] of longint;
  m, w, l, mleap, dleap, yul, mul, dul: longint;
  a, b, n, sum, i, j, dow, max, maxcount: longint;
begin
  read(m, w, l, mleap, dleap);
  sum := 0;
  for i := 1 to m do
  begin
    read(months[i]);
    sum := sum + months[i];
  end;
  read(yul, mul, dul, a, b);

  // first we assume the day before 1st Jan, 1 is day 0
  dow := 0;
  for i := 1 to 1000000 do
  begin
    dows[i] := dow;
    dow := dow + sum;
    if (i mod l = 0) then
      dow := dow + dleap;
    dow := dow mod w;
  end;

  // dow of given date
  n := dows[yul];
  for i := 1 to mul - 1 do
  begin
    n := n + months[i];
    if ((yul mod l = 0) and (i = mleap)) then
      n := n + dleap;
  end;
  n := (n + dul) mod w;

  // simulate all different day of week on day before 1st Jan
  for i := 0 to w - 1 do
  begin
    noleap[i] := 0;
    dow := i;
    for j := 1 to m do
    begin
      if ((months[j] >= 13) and ((dow + 13) mod w = n)) then
        inc(noleap[i]);
      dow := dow + months[j];
    end;
  end;
  for i := 0 to w - 1 do
  begin
    leap[i] := 0;
    dow := i;
    for j := 1 to m do
    begin
      if ((months[j] >= 13) or ((j = mleap) and (months[j] + dleap >= 13))) then
        if ((dow + 13) mod w = n) then
          inc(leap[i]);
      dow := dow + months[j];
      if (j = mleap) then
        dow := dow + dleap;
    end;
  end;

  // lastly scan through year a to b to find max
  max := 0;
  maxcount := 0;
  for i := a to b do
  begin
    if (i mod l = 0) then
    begin
      if (leap[dows[i]] > max) then
      begin
        max := leap[dows[i]];
        maxyears[1] := i;
        maxcount := 1;
      end
      else if (leap[dows[i]] = max) then
      begin
        maxyears[maxcount + 1] := i;
        inc(maxcount);
      end
    end
    else if (noleap[dows[i]] > max) then
    begin
      max := noleap[dows[i]];
      maxyears[1] := i;
      maxcount := 1;
    end
    else if (noleap[dows[i]] = max) then
    begin
      maxyears[maxcount + 1] := i;
      inc(maxcount);
    end;
  end;

  //output
  writeln(max);
  for i := 1 to maxcount - 1 do
    write(maxyears[i], ' ');
  writeln(maxyears[maxcount])
end.